File: mojo/src/main/java/com/antwerkz/sofia/SofiaMojo.java
package com.antwerkz.sofia;
/*
* Copyright 2001-2005 The Apache Software Foundation.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*      http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import java.io.File;
import java.io.IOException;
import java.nio.charset.Charset;

import org.apache.maven.plugin.AbstractMojo;
import org.apache.maven.plugin.MojoExecutionException;
import org.apache.maven.plugins.annotations.LifecyclePhase;
import org.apache.maven.plugins.annotations.Mojo;
import org.apache.maven.plugins.annotations.Parameter;
import org.apache.maven.project.MavenProject;

@Mojo(name = "generate", defaultPhase = LifecyclePhase.GENERATE_SOURCES)
public class SofiaMojo extends AbstractMojo {
  @Parameter(property = "project")
  private MavenProject project;
  @Parameter(property = "outputDirectory", defaultValue = "${project.build.directory}/generated-sources/sofia")
  private File outputDirectory;
  @Parameter(defaultValue = "src/main/resources/sofia.properties")
  private File inputFile;
  @Parameter(defaultValue = "com.antwerkz.sofia")
  private String packageName;
  @Parameter(defaultValue = "false")
  private boolean playController;
  @Parameter(defaultValue = "jul")
  private String loggingType;
  @Parameter(defaultValue = "src/main/webapp/js/sofia.js")
  private File jsOutputFile;
  @Parameter(defaultValue = "false")
  private boolean javascript;
  @Parameter(defaultValue = "ISO-8859-1")
  private String charset;

  public void execute() throws MojoExecutionException {
    if (!outputDirectory.exists()) {
      outputDirectory.mkdirs();
    }
    try {
      generate();
      project.addCompileSourceRoot(outputDirectory.getAbsolutePath());
    } catch (Exception e) {
      throw new MojoExecutionException(e.getMessage(), e);
    }
  }

  public void generate() throws IOException, MojoExecutionException {
    if (!inputFile.exists()) {
      throw new MojoExecutionException("Missing inputFile file: " + inputFile);
    }
    new LocalizerGenerator(new SofiaConfig()
      .setBundleName(inputFile.getName())
      .setPackageName(packageName)
      .setProperties(inputFile)
      .setType(loadLoggingType())
      .setUseControl(playController)
      .setOutputDirectory(outputDirectory)
      .setGenerateJavascript(javascript)
      .setJavascriptOutputFile(jsOutputFile)
      .setCharset(Charset.forName(charset))
    ).write();
  }

  private LoggingType loadLoggingType() throws MojoExecutionException {
    try {
      return loggingType == null ? null : LoggingType.valueOf(loggingType.toUpperCase());
    } catch (IllegalArgumentException e) {
      throw new MojoExecutionException("Unknown logging type: " + loggingType);
    }
  }

  public File getInputFile() {
    return inputFile;
  }

  @Override
  public String toString() {
    final StringBuilder sb = new StringBuilder("SofiaMojo{");
    sb.append("outputDirectory=").append(outputDirectory);
    sb.append(", inputFile=").append(inputFile);
    sb.append(", packageName='").append(packageName).append('\'');
    sb.append(", playController=").append(playController);
    sb.append(", loggingType='").append(loggingType).append('\'');
    sb.append(", jsOutputFile=").append(jsOutputFile);
    sb.append(", javascript=").append(javascript);
    sb.append(", charset='").append(charset).append('\'');
    sb.append('}');
    return sb.toString();
  }
}

File: mojo/src/main/kotlin/com/antwerkz/sofia/SofiaMojo.kt
package com.antwerkz.sofia

import org.apache.maven.plugin.AbstractMojo
import org.apache.maven.plugin.MojoExecutionException
import org.apache.maven.plugins.annotations.LifecyclePhase
import org.apache.maven.plugins.annotations.Mojo
import org.apache.maven.plugins.annotations.Parameter
import org.apache.maven.project.MavenProject
import java.io.File
import java.nio.charset.Charset
import java.util.Locale

@Mojo(name = "generate", defaultPhase = LifecyclePhase.GENERATE_SOURCES)
open class SofiaMojo : AbstractMojo() {

    @Parameter(property = "project")
    lateinit var project: MavenProject
    @Parameter(property = "outputDirectory", defaultValue = "\${project.build.directory}/generated-sources/sofia")
    lateinit var outputDirectory: File
    @Parameter(defaultValue = "src/main/resources/sofia.properties")
    lateinit var inputFile: File
    @Parameter(defaultValue = "com.antwerkz.sofia")
    lateinit var packageName: String
    @Parameter
    var className: String? = null
    @Parameter(defaultValue = "slf4j")
    lateinit var loggingType: String
    @Parameter(defaultValue = "java")
    lateinit var outputType: String
//    @Parameter(defaultValue = "src/main/webapp/js/sofia.js")
//    lateinit var jsOutputFile: File
//    @Parameter(defaultValue = "false")
//    var javascript: Boolean = false
    @Parameter(defaultValue = "ISO-8859-1")
    lateinit var charset: String

    override fun execute() {
        if (!outputDirectory.exists()) {
            outputDirectory.mkdirs()
        }
        try {
            generate()
            project.addCompileSourceRoot(outputDirectory.absolutePath)
        } catch (e: Exception) {
            throw MojoExecutionException(e.message, e)
        }
    }

    fun generate() {
        if (!inputFile.exists()) {
            throw MojoExecutionException("Missing inputFile file: " + inputFile)
        }
        val config = SofiaConfig(inputFile, outputDirectory, loadLoggingType(), packageName,
                generateJava = this.outputType == "java", generateKotlin = this.outputType == "kotlin",
                className = this.className ?: inputFile.nameWithoutExtension.capitalizeFirstLetter(), charset = Charset.forName(charset))
        LocalizerGenerator(config).write()
    }

    private fun loadLoggingType(): LoggingType {
        try {
            return LoggingType.valueOf(loggingType.uppercase(Locale.getDefault()))
        } catch (e: IllegalArgumentException) {
            throw MojoExecutionException("Unknown logging type: " + loggingType)
        }

    }

    override fun toString(): String {
        return "SofiaMojo(outputDirectory=$outputDirectory, inputFile=$inputFile, packageName='$packageName'," +
                " className=$className, loggingType='$loggingType', outputType='$outputType', charset='$charset')"
    }
}
